File: 2026/08-August/13-Thursday/solution.java
class Solution {

    // Node:
    // leftChar, rightChar, prefix, suffix, best, length
    static class Node {
        char leftChar;
        char rightChar;
        int prefix;
        int suffix;
        int best;
        int length;

        Node(char ch) {
            leftChar = ch;
            rightChar = ch;
            prefix = 1;
            suffix = 1;
            best = 1;
            length = 1;
        }
    }

    Node[] tree;

    public int[] longestRepeating(String s, String queryCharacters, int[] queryIndices) {
        int n = s.length();

        tree = new Node[4 * n];

        build(s, 1, 0, n - 1);

        int[] answer = new int[queryIndices.length];

        for (int i = 0; i < queryIndices.length; i++) {
            update(
                1,
                0,
                n - 1,
                queryIndices[i],
                queryCharacters.charAt(i)
            );

            answer[i] = tree[1].best;
        }

        return answer;
    }

    private void build(String s, int node, int left, int right) {
        if (left == right) {
            tree[node] = new Node(s.charAt(left));
            return;
        }

        int mid = left + (right - left) / 2;

        build(s, node * 2, left, mid);
        build(s, node * 2 + 1, mid + 1, right);

        tree[node] = merge(tree[node * 2], tree[node * 2 + 1]);
    }

    private void update(
        int node,
        int left,
        int right,
        int index,
        char ch
    ) {
        if (left == right) {
            tree[node] = new Node(ch);
            return;
        }

        int mid = left + (right - left) / 2;

        if (index <= mid) {
            update(node * 2, left, mid, index, ch);
        } else {
            update(node * 2 + 1, mid + 1, right, index, ch);
        }

        tree[node] = merge(tree[node * 2], tree[node * 2 + 1]);
    }

    private Node merge(Node a, Node b) {
        Node res = new Node(a.leftChar);

        res.leftChar = a.leftChar;
        res.rightChar = b.rightChar;
        res.length = a.length + b.length;

        // Initially take the best from either side
        res.best = Math.max(a.best, b.best);

        res.prefix = a.prefix;
        res.suffix = b.suffix;

        // Characters at the boundary are equal
        if (a.rightChar == b.leftChar) {

            // Combine suffix of left segment
            // with prefix of right segment
            res.best = Math.max(
                res.best,
                a.suffix + b.prefix
            );

            // If the entire left segment is one character
            if (a.prefix == a.length) {
                res.prefix = a.length + b.prefix;
            }

            // If the entire right segment is one character
            if (b.suffix == b.length) {
                res.suffix = b.length + a.suffix;
            }
        }

        return res;
    }
}

File: 2026/08-August/13-Thursday/solution.py
class Solution:
    def longestRepeating(self, s: str, queryCharacters: str, queryIndices: list[int]) -> list[int]:
        n = len(s)

        # Each node:
        # [left_char, right_char, prefix, suffix, best]
        tree = [None] * (4 * n)

        def make_node(ch):
            return [ch, ch, 1, 1, 1]

        def merge(a, b):
            if a is None:
                return b
            if b is None:
                return a

            left = a[0]
            right = b[1]

            pref = a[2]
            suff = b[3]
            best = max(a[4], b[4])

            # Entire boundary has the same character
            if a[1] == b[0]:
                best = max(best, a[3] + b[2])

                if a[2] == a[3]:
                    pref = a[2] + b[2]

                if b[2] == b[3]:
                    suff = b[3] + a[3]

            return [left, right, pref, suff, best]

        def build(node, l, r):
            if l == r:
                tree[node] = make_node(s[l])
                return

            mid = (l + r) // 2
            build(node * 2, l, mid)
            build(node * 2 + 1, mid + 1, r)
            tree[node] = merge(tree[node * 2], tree[node * 2 + 1])

        def update(node, l, r, idx, ch):
            if l == r:
                tree[node] = make_node(ch)
                return

            mid = (l + r) // 2

            if idx <= mid:
                update(node * 2, l, mid, idx, ch)
            else:
                update(node * 2 + 1, mid + 1, r, idx, ch)

            tree[node] = merge(tree[node * 2], tree[node * 2 + 1])

        build(1, 0, n - 1)

        ans = []

        for ch, idx in zip(queryCharacters, queryIndices):
            update(1, 0, n - 1, idx, ch)
            ans.append(tree[1][4])

        return ans
